catkit2/testbed/tracing.py

```diff
@@ -0,0 +1,218 @@
+
+    # Write JSON to file.
+    f.write(data + ',\n')
+
+
+class TraceWriter:
+    '''A writer for performance trace logs.
+
+    Parameters
+    ----------
+    host : str
+        The host which distributes the trace messages.
+    port : int
+        The port on which the host distributes the trace messages.
+    '''
+    def __init__(self, host, port):
+        self.f = None
+
+        self.context = zmq.Context()
+
+        self.host = host
+        self.port = port
+
+        self.shutdown_flag = threading.Event()
+        self.thread = None
+
+    def open(self, filename):
+        '''Open the writer.
+
+        Parameters
+        ----------
+        filename : str
+            The path to the file where to write the performance trace.
+
+        Returns
+        -------
+        TraceWriter
+            The current trace writer. This is for use as a context manager.
+        '''
+        self.shutdown_flag.clear()
+
+        self._filename = filename
+
+        self.thread = threading.Thread(target=self._loop)
+        self.thread.start()
+
+        return self
+
+    def close(self):
+        '''Close the writer.
+        '''
+        self.shutdown_flag.set()
+
+        if self.thread:
+            self.thread.join()
+
+    def _loop(self):
+        # Set up socket.
+        socket = self.context.socket(zmq.SUB)
+        socket.connect(f'tcp://{self.host}:{self.port}')
+        socket.subscribe('')
+        socket.RCVTIMEO = 50
+
+        # Set up cache for process names.
+        process_names = {}
+        thread_names = {}
+
+        # Set initial time.
+        # Subtracting one second before current time to hopefully never have negative timestamps.
+        t_0 = catkit_bindings.get_timestamp() - 1_000_000_000
+
+        with open(self._filename, 'w') as f:
+            # Write JSON header.
+            f.write('[\n')
+
+            # Main loop.
+            while not self.shutdown_flag.is_set():
+                # Receive a new trace event.
+                try:
+                    message = socket.recv_multipart()
+                except zmq.ZMQError as e:
+                    if e.errno == zmq.EAGAIN:
+                        # Timed out.
+                        continue
+                    else:
+                        raise RuntimeError('Error during receive.') from e
+
+                # Decode event.
+                proto_event = tracing_proto.TraceEvent()
+                proto_event.ParseFromString(message[0])
+
+                # Convert event to JSON format.
+                event_type = proto_event.WhichOneof('event')
+                if event_type == 'interval':
+                    raw_data = proto_event.interval
+                    data = {
+                        'name': raw_data.name,
+                        'cat': raw_data.category,
+                        'ph': 'X',
+                        'ts': (raw_data.timestamp - t_0) / 1000,
+                        'dur': raw_data.duration / 1000,
```

Comment:
Should there be a note somewhere about the units of the time stamp and durations here or is this obvious somewhere else? Same for all other instances in here that have a hard-coded unit conversion.

Reply:
Will do. There are hardcoded units everywhere, and they are fixed by the Chrome tracing format.

Reply:
Done.